### ISSUE TITLE
STARsolo for multi-modal data

### ISSUE THREAD

**Issue by andrewjkwok** [alexdobin/STAR#1235]
Hi,

Really delighted with this incredible set of software now with single cell capacities. I wanted to ask about whether any attempts have been made to adapt the STARsolo pipeline for cell surface protein profiling. From the documentation and the preprint, it seems that STARsolo doesn't currently support this yet. 

I was wondering if it makes sense (or maybe has been attempted already) to just add the antibody oligo sequences to the genome fasta file to then create the index and directly map to this?

Thanks in advance.

**Comment by alexdobin** [comment]
Hi Andrew,

I have promised to add this capability for a while, and now I am starting to look into it.
I think simply adding the sequences to the genome as separate "chromosomes" should work.
You would also need to add the corresponding exon lines (with transcript_id and gene_id) into the GTF file.
The added sequences should contain the barcodes and 5' and 3' adapters.

One problem with this approach is that you cannot enforce a higher "alignment quality" for mapping to these barcodes - e.g. requiring end-to-end alignment with fewer mismatches.

Cheers
Alex

**Comment by andrewjkwok** [comment]
Hi Alex,

Thanks for the reply - looking forward to when the functionality gets released. I'll give the manual addition a shot, but wanted to just quickly clarify about the added sequences containing the barcodes and adapters - what exactly do you mean by that? 

My read structure (experiment done on the BD Rhapsody platform) is such that only read 2 contains the cDNA information (similar to 10X - I understand STARsolo requires putting R2 as the first file of a pair), so surely I only need to add 1 adapter downstream to the cDNA. The second thing is - for the UMI/cell barcode, will the mapping accuracy be much reduced if just put all Ns in these positions?

**Comment by alexdobin** [comment]
Hi Andrew,

could you explain in detail how the read1 and read2 with the antibody barcodes look like?

Cheers
Alex

**Comment by andrewjkwok** [comment]
The read structure is like this: https://teichlab.github.io/scg_lib_structs/methods_html/BD_Rhapsody.html

Read 1 starts with a 52 bp cell barcode followed by an 8bp UMI, then a run of Ts for the polyA capture (should be 18Ts). Read 2 contains the cDNA information and begins directly with the cDNA - either from the RNA or antibodies. The antibody oligos are 36bp long.

This dataset was run 150bp PE, so for read2, I imagine for the antibodies it would be 36bp of the antibody sequence, followed by 18Ts, followed by the 8bp UMI, followed by the 52bp cell barcode and then finally 36bp of Illumina Truseq Read 1 sequence / the Illumina P5 sequence (36 + 18 + 8 + 52 + 36 = 150).

**Comment by alexdobin** [comment]
Hi Andrew,

since only 36 first bases of the barcode, read are informative, it's best to clip out the rest of the bases with 
by trimming on polyA stretch.
You can probably write a simple script to trim the reads that have 18As starting from position 37.
You can also try to use STAR's internal trimmer with
*--clip3pAdapterSeq AAAAAAAAAAAAAAAAAA*  (18 As).
Note that STAR will clip off this sequence no matter where it's located along the read, so some of the cDNA reads with long A stretches will get trimmed. Also, any As at the ends of cDNA will get trimmed.
You can play with the length of this polyA sequence, reducing it will make trimming more permissive.
Another parameter to play with is *--clip3pAdapterMMp 0.1* (default) which determines the max proportion of bases that can be mismatched, for 18b this would allow 1 mismatch.

At the genome generation step, you will need to add only the 36b barcode sequences are "chromosomes" to FASTA and genes to GTF.

Cheers
Alex

**Comment by andrewjkwok** [comment]
Thanks Alex - that's very helpful. Looking at issue #1238 I was wondering whether you would recommend what you suggested above in terms of incorporating all my antibody sequences into a single reference, or the solution that Tatyana proposed in the issue of creating a separate antibody only reference and using the unmapped reads against that? 

I presume if I end up generating more data where my antibody and gene expression libraries are separate, I would map my antibody libraries to a specific antibody only reference as mentioned in issue #1238 and then recombine with the gene expression downstream - in that case, are there recommendations in terms of various options that might be worth noting? 

Many thanks in advance,
Andrew

**Comment by andrewjkwok** [comment]
Hi Alex,

I have an update on the antibody read structure on the BD Rhapsody platform and it isn't as simple as I'd hoped previously. It seems there is actually an extra 12bp UMI at the beginning of the cDNA read (read 2, which is fed into STARsolo as the first file). Is there any flexibility in STARsolo to put multiple UMIs for the _--soloUMIposition_ option?

**Comment by alexdobin** [comment]
Hi Andrew,

No, presently, multiple UMIs are not supported.

Looking at the Teichmann lab description, there are 3 CBs, each 9b long, right?

A workaround would be to preprocess reads to combine UMI sequences into one sequence: cut 12b from the cDNA read and add it to the barcode read. To do this, you would need to decide whether the read is RNA or antibody - by checking the presence of 18As starting from position 49.
You would have to have longer UMIs for RNA reads as well - you can attach a stretch of extra 8 As
Note that up to 16b UMI are presently supported, so you can use only 16b out of 8+12b for UMIs.

The UMI-8b/CB-52b on the cDNA read are the same as on the barcode read, right?
Then you do not really need them.

To summarize the preprocessing (read1=barcode, read2=cDNA/antibody):
1. Check read2 for the presence of 18A starting from base 49 (maybe allow for 1 or 2 mismatches).
2. No:  add 8A to the end of the read. Record the full cDNA read as read2.
3. Yes: add 8b from the antibody read UMI to the end of read1. Record the 36b antibody sequence as read2.

Cheers
Alex

**Comment by apredeus** [comment]
I'd be also really interested in helping with testing/developing the capabilities for CITE-seq/cell hashing. Right now we moved pretty much all of the public and internal data re-processing to `STARsolo`, and the only important and popular thing that's missing is this.

**Comment by alexdobin** [comment]
I  did promise CITE-seq processing a long time ago... but still not sure when I will be able to work on it. :(
There are some hacks that can be used, such as representing the tags as separate "chromosomes".
